### PR TITLE
[p2p] Add libp2p_metrics into host and use bwCounter to record in/out…

### PR DESCRIPTION
**Pull request by jz0000**
… networking traffic in a node

## Issue

So far we don't have good visibility into what's going on in the p2p layer, to improve the performance of p2p layer, we need to have some metrics to be collected.  This is the first step, I will gradually add more metrics in more places as needed. 

The changes in the go.mod is automatically generated when I ran "make" in my workspace, I have retried a few times and it seems not harmful.  Please let me know if otherwise. Thanks.

### Unit Test Coverage

I have run go test -cover inside node directory, let me know if I need to run some other unit testing. 

Jians-MacBook-Pro:node jz$ go test -cover
PASS
coverage: 12.4% of statements
ok  	github.com/harmony-one/harmony/node	1.195s

### Test/Run Logs

Have run ./test/debug.sh and generated logs  like 

tmp_log/log-20200419-232519/zerolog-validator-127.0.0.1-9000.log, on my machine. Here I share a few examples of the generated messages in the log:

https://gist.github.com/jz0000/807652e4b0f05b91eae3319246652f6e

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)

NO

2. **Describe the migration plan.**. For each flag epoch, describe what changes take place at the flag epoch, the anticipated interactions between upgraded/non-upgraded nodes, and any special operational considerations for the migration.

3. **Describe how the plan was tested.**

4. **How much minimum baking period after the last flag epoch should we allow on Pangaea before promotion onto mainnet?**

5. **What are the planned flag epoch numbers and their ETAs on Pangaea?**

6. **What are the planned flag epoch numbers and their ETAs on mainnet?**

    Note that this must be enough to cover baking period on Pangaea.

7. **What should node operators know about this planned change?**

8. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)

NO

9. **Does the existing `node.sh` continue to work with this change?**

10. **What should node operators know about this change?**

11. **Does this PR introduce significant changes to the operational requirements of the node software, such as >20% increase in CPU, memory, and/or disk usage?**

There could be some minor increase in the log it generated, thus disk usage might increase a bit, but don't see this an issue.

### CODE REVIEW

File: p2p/host.go [hunk 1]
```diff
@@ -88,14 +94,18 @@ func NewHost(self *Peer, key libp2p_crypto.PrivKey) (Host, error) {
 
 	self.PeerID = p2pHost.ID()
 	subLogger := utils.Logger().With().Str("hostID", p2pHost.ID().Pretty()).Logger()
+
+	newMetrics := libp2p_metrics.NewBandwidthCounter()
```
Comment (denniswon) [inline p2p/host.go:98]: make this as an option param to node.sh? Might want to avoid having "all nodes" logging this metric.

**Comment by fxfactorial** [comment]
Please let’s hold off on this until after my bigger PR finished

**Comment by LeoHChen** [comment]
Jenkins, test this please.

File: p2p/host.go (hunk 1 above)
Comment (LeoHChen) [reply p2p/host.go:98]: this is local log, should be fine.

File: p2p/host.go [hunk 2]
```diff
@@ -161,7 +174,14 @@ func (host *HostV2) SendMessageToGroups(groups []nodeconfig.GroupID, msg []byte)
 			err = e
 			continue
 		}
+		// log out-going metrics
+		host.metrics.LogSentMessage(int64(len(msg)))
 	}
+	host.logger.Info().
+		Int64("TotalOut", host.GetBandwidthTotals().TotalOut).
+		Float64("RateOut", host.GetBandwidthTotals().RateOut).
+		Msg("Record Sending Metrics!")
```
Comment (LeoHChen) [inline p2p/host.go:183]: use the same prefix for metrics

File: p2p/host.go (hunk 1 above)
Comment (jz0000) [reply p2p/host.go:98]: ok.

File: node/node_handler.go [hunk 3]
```diff
@@ -448,6 +455,10 @@ func (node *Node) PostConsensusProcessing(
 	// Broadcast client requested missing cross shard receipts if there is any
 	node.BroadcastMissingCXReceipts()
 
+	// Clear metrics after one consensus cycle
+	node.host.ResetMetrics()
+	utils.Logger().Info().Msg("Metrics cleared after one consensus cycle")
```
[the remark this answers is not in the document]
Comment (jz0000) [reply node/node_handler.go:460]: Unfortunately this is needed. Our data scientist will need to know when this counter resets, to be able to compute the sum correctly, if needed. 

Thought about removing the reset all together, but that might let counters wrap around after sometime, once the network becomes bigger.

File: node/node_handler.go [hunk 4]
```diff
@@ -59,6 +59,13 @@ func (node *Node) processSkippedMsgTypeByteValue(
 
 // HandleMessage parses the message and dispatch the actions.
 func (node *Node) HandleMessage(content []byte, sender libp2p_peer.ID) {
+	// log in-coming metrics
+	node.host.LogRecvMessage(content)
+	utils.Logger().Info().
+		Int64("TotalIn", node.host.GetBandwidthTotals().TotalIn).
+		Float64("RateIn", node.host.GetBandwidthTotals().RateIn).
+		Msg("Record Receiving Metrics!")
```
[the remark this answers is not in the document]
Comment (jz0000) [reply node/node_handler.go:67]: Sounds good.

File: p2p/host.go (hunk 2 above)
Comment (jz0000) [reply p2p/host.go:183]: ok.

**Comment by fxfactorial** [comment]
@jz0000 as I already wrote : let’s hold off on this PR because it will cause huge and needless merge conflicts when al this code is changing anyway under my p2p revamp

**Comment by jz0000** [comment]
> @jz0000 as I already wrote : let’s hold off on this PR because it will cause huge and needless merge conflicts when al this code is changing anyway under my p2p revamp

Leo let me to merge to s3 branch, you are going to master correct?

**Comment by fxfactorial** [comment]
> > @jz0000 as I already wrote : let’s hold off on this PR because it will cause huge and needless merge conflicts when al this code is changing anyway under my p2p revamp
> 
> Leo let me to merge to s3 branch, you are going to master correct?

ah, my bad - yes, I'm targetting master

**Comment by jz0000** [comment]
> > > @jz0000 as I already wrote : let’s hold off on this PR because it will cause huge and needless merge conflicts when al this code is changing anyway under my p2p revamp
> > 
> > 
> > Leo let me to merge to s3 branch, you are going to master correct?
> 
> ah, my bad - yes, I'm targetting master

**Pull request merged**